### PR TITLE
array_like_handler wrapper

### PR DESCRIPTION
**Pull request by Ishticode**

### CODE REVIEW

File: ivy/func_wrapper.py [hunk 1]
```diff
@@ -45,6 +47,34 @@ def _get_first_array(*args, **kwargs):
 # ---------------#
 
 
+def handle_array_like(fn: Callable) -> Callable:
+
+    @functools.wraps(fn)
+    def new_fn(*args, **kwargs):
+        args = list(args)
+        parameters = list(typing.get_type_hints(fn))
```
Comment (djl11) [inline ivy/func_wrapper.py:55]: calling `typing.get_type_hints(fn)` twice is very inefficient. You should cache this first: `type_hints = typing.get_type_hints(fn)`

File: ivy/func_wrapper.py [hunk 2]
```diff
@@ -45,6 +47,34 @@ def _get_first_array(*args, **kwargs):
...
+        annotations = list(typing.get_type_hints(fn).values())
+
+        # accessing common index across input arg values, parameters and annotations.
+        for i in range(len(annotations)):
```
Comment (djl11) [inline ivy/func_wrapper.py:59]: `for i, (annotation, parameter, arg) in enumerate(zip(annotations, parameters, args)):`

File: ivy/func_wrapper.py [hunk 3]
```diff
@@ -45,6 +47,34 @@ def _get_first_array(*args, **kwargs):
...
+
+            if "Array" in str(annotations[i]) and all(
```
Comment (djl11) [inline ivy/func_wrapper.py:61]: `if "Array" in str(annotation) and all(`

etc.

File: ivy/func_wrapper.py [hunk 4]
```diff
@@ -45,6 +47,34 @@ def _get_first_array(*args, **kwargs):
...
+                    sq not in str(annotations[i])
+                    for sq in ["Sequence", "List", "Tuple"]):
+
+                if i < len(args):
```
Comment (djl11) [inline ivy/func_wrapper.py:65]: `num_args = len(args)` should be cached outside the for loop

File: ivy/func_wrapper.py [hunk 5]
```diff
@@ -45,6 +47,34 @@ def _get_first_array(*args, **kwargs):
...
+
+    @functools.wraps(fn)
+    def new_fn(*args, **kwargs):
+        args = list(args)
+        parameters = list(typing.get_type_hints(fn))
+        annotations = list(typing.get_type_hints(fn).values())
+
+        # accessing common index across input arg values, parameters and annotations.
+        for i in range(len(annotations)):
+
```
Comment (djl11) [inline ivy/func_wrapper.py:60]: `annotation_str = str(annotation)` should be cached

File: ivy/func_wrapper.py [hunk 6]
```diff
@@ -45,6 +47,34 @@ def _get_first_array(*args, **kwargs):
...
+            if "Array" in str(annotations[i]) and all(
+                    sq not in str(annotations[i])
+                    for sq in ["Sequence", "List", "Tuple"]):
+
+                if i < len(args):
+                    if isinstance(args[i], (list, tuple)):
```
Comment (djl11) [inline ivy/func_wrapper.py:66]: `if isinstance(arg, (list, tuple)):`

etc.

File: ivy/func_wrapper.py [hunk 7]
```diff
@@ -45,6 +47,34 @@ def _get_first_array(*args, **kwargs):
...
+                        args[i] = ivy.array(args[i])
```
Comment (djl11) [inline ivy/func_wrapper.py:67]: `args[i] = ivy.array(arg)`

File: ivy/func_wrapper.py [hunk 8]
```diff
@@ -45,6 +47,34 @@ def _get_first_array(*args, **kwargs):
...
+                elif parameters[i] in kwargs:
```
Comment (djl11) [inline ivy/func_wrapper.py:68]:
```
elif parameter in kwargs:
    kwarg = kwargs[parameter]
    if isinstance(kwarg, (list, tuple)):
        kwargs[parameter] = ivy.array(kwarg)
```

**Review by djl11 — CHANGES_REQUESTED** [review]
general feedback, cache what you can, don't compute the same thing repeatedly. If something is called more than once in the script, do it once and store it rather than continuing to compute it repeatedly :)

**Comment by Ishticode** [comment]
Thanks Dan. Very useful principal indeed and I am not sure why i didn't use zip in the first place 🙃.

**Review by djl11 — APPROVED** [review]
lgtm!

**Pull request merged**